### PR TITLE
Various improvements/additions/fixes

### PR DESCRIPTION
**Pull request by akemin-dayo**
- `dm.pl` now allows users to use package IDs that contain uppercase characters
- Running `make package install` directly on a device as `mobile` is now supported, as `install.exec` now uses `su` as opposed to `sh`
- Added empty `sdks/` and `toolchains/` folders
- Ignore deprecation messages generated by recent versions of Perl when executing `aliased.pm` (temporary workaround, a real solution would be to figure out what exactly was deprecated and fix this in  a backwards-compatible manner)
- Added/fixed full Linux support
- Fixed `lipo` support on all non-OS X platforms
- Remove `gcc`/`g++`
- Fixed a bug where makedeps would always be activated due a logic error in `rules.mk`
- Updated all targets against OS X's copy
- Added an user-definable `TARGET_INSTALL_REMOTE` variable to control whether or not Theos should install via `ssh` on an iOS device (for those who want to do password-less SSH key authentication from `mobile`)
- Added missing rules for all `PRINT_FORMAT_*` calls
- Improved and unified SDK selection logic
- Improved and unified minimum OS flag selection logic

Fully tested on OS X, iOS, Windows, and Linux against all my projects.

### CODE REVIEW

**Comment by Razzile** [comment]
:+1: nice one

File: .gitignore [hunk 1]
```diff
@@ -2,8 +2,8 @@
 .DS_Store
 ._*
 
-sdks
-toolchain
+sdks/*.sdk
```
Comment (kirb) [inline .gitignore]: I figure this should just be `*`, not specifically `*.sdk`.

File: bin/install.exec [hunk 2]
```diff
@@ -15,5 +15,5 @@ if [[ TARGET_INSTALL_REMOTE -eq 1 ]]; then
 
 	exec ${args[@]}
 else
-	exec sh -c "$@"
+	exec su -c "$@"
```
Comment (kirb) [inline bin/install.exec:18]: Not sure about this. It should be ok, but there may be some edge cases of projects that assume their `install.exec` runs as the same user.

File: makefiles/targets/Darwin-arm/iphone.mk [hunk 3]
```diff
@@ -106,11 +126,16 @@ else
...
-_THEOS_TARGET_CFLAGS := $(SDKFLAGS)
-_THEOS_TARGET_LDFLAGS := $(SDKFLAGS) $(LEGACYFLAGS) -multiply_defined suppress
+SDKFLAGS := -D__IPHONE_OS_VERSION_MIN_REQUIRED=__IPHONE_$(subst .,_,$(_THEOS_TARGET_IPHONEOS_DEPLOYMENT_VERSION))
+VERSIONFLAGS := -miphoneos-version-min=$(_THEOS_TARGET_IPHONEOS_DEPLOYMENT_VERSION)
+
+_THEOS_TARGET_CFLAGS := -isysroot "$(ISYSROOT)" $(SDKFLAGS) $(VERSIONFLAGS) $(_THEOS_TARGET_CC_CFLAGS) $(MODULESFLAGS)
+_THEOS_TARGET_LDFLAGS := -isysroot "$(SYSROOT)" $(SDKFLAGS) $(VERSIONFLAGS) $(LEGACYFLAGS) -multiply_defined suppress
 
 TARGET_INSTALL_REMOTE := $(_THEOS_FALSE)
+ifeq ($(THEOS_INSTALL_USING_SSH),1)
```
Comment (kirb) [inline makefiles/targets/Darwin-arm/iphone.mk]: This should use `$(call __theos_bool)`.

File: makefiles/targets/Darwin-arm/iphone.mk [hunk 4]
```diff
@@ -2,24 +2,25 @@ ifeq ($(_THEOS_TARGET_LOADED),)
...
+_SDKVERSION := $(or $(__THEOS_TARGET_ARG_$(word 1,$(_THEOS_TARGET_ARG_ORDER))),$(SDKVERSION_$(THEOS_CURRENT_ARCH)),$(SDKVERSION))
 _THEOS_TARGET_SDK_VERSION := $(or $(_SDKVERSION),latest)
+_THEOS_TARGET_INCLUDE_SDK_VERSION := $(or $(INCLUDE_SDKVERSION),$(INCLUDE_SDKVERSION_$(THEOS_CURRENT_ARCH)),latest)
 
 _SDK_DIR := $(THEOS)/sdks
 _IOS_SDKS := $(sort $(patsubst $(_SDK_DIR)/iPhoneOS%.sdk,%,$(wildcard $(_SDK_DIR)/iPhoneOS*.sdk)))
 ifeq ($(words $(_IOS_SDKS)),0)
-@$(PRINT_FORMAT_ERROR) "You do not have an SDK in $(_SDK_DIR)." >&2
+before-all::
+	@$(PRINT_FORMAT_ERROR) "You do not have an SDK in $(_SDK_DIR)." >&2
```
Comment (kirb) [inline makefiles/targets/Darwin-arm/iphone.mk]: All instances of error cases should also `exit 1` afterwards.

File: bin/install.exec (hunk 2 above)
Comment (akemin-dayo) [inline bin/install.exec:18]: Well, most users' usage of `install.exec` involves connecting to their device via SSH as `root` anyway, so... I think this change should be fine?

Comment (kirb) [inline bin/install.exec:18]: Fair.

File: makefiles/targets/Darwin-arm/iphone.mk (hunk 4 above)
Comment (akemin-dayo) [inline makefiles/targets/Darwin-arm/iphone.mk]: Fixed in `0f4b39df56018cb6b97e0a921bfb01fd491349d0`

File: makefiles/targets/Darwin-arm/iphone.mk (hunk 3 above)
Comment (akemin-dayo) [inline makefiles/targets/Darwin-arm/iphone.mk]: Fixed in `91ad1c404927a412d56d60cc6288846fec8ce4f5`

File: .gitignore (hunk 1 above)
Comment (akemin-dayo) [inline .gitignore]: Fixed in `4867dcbad8e1b6d41b9e85825134a36073911845`

**Comment by kirb** [comment]
Rebased commits to change commit messages and merged them as https://github.com/kirb/theos/compare/372431c08836...759a29fb3856. Cheers 👍

**Pull request closed without merging**